cmd/kubeapps-apis/server/packages.go

```diff
@@ -0,0 +1,82 @@
+/*
+Copyright © 2021 VMware
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package server
+
+import (
+	"context"
+
+	packages "github.com/kubeapps/kubeapps/cmd/kubeapps-apis/gen/core/packages/v1alpha1"
+)
+
+// packagesServer implements the API defined in proto/kubeappsapis/core/packages/v1alpha1/packages.proto
+type packagesServer struct {
+	packages.UnimplementedPackagesServiceServer
+
+	// plugins is a slice of all registered plugins which satisfy the core.packages.v1alpha1
+	// interface.
+	plugins []*pkgsPluginWithServer
+}
+
+func NewPackagesServer(plugins []*pkgsPluginWithServer) *packagesServer {
+	return &packagesServer{
+		plugins: plugins,
+	}
+}
+
+// GetAvailablePackages returns the packages based on the request.
+func (s packagesServer) GetAvailablePackages(ctx context.Context, request *packages.GetAvailablePackagesRequest) (*packages.GetAvailablePackagesResponse, error) {
+	pkgs := []*packages.AvailablePackage{}
+	for _, p := range s.plugins {
+		response, err := p.server.GetAvailablePackages(ctx, request)
```

Comment:
So there were a number of different ways to call the separate plugin versions. The simplest would have been to use the client and actually do the grpc call to localhost, but that then requires serializing/deserializing for every request. Instead, because we recorded at startup which plugins satisfy the core packaging server interface and store a reference to the implementation, we can just use the implementations directly here without going over the wire at all, so no serialization/deserialization required.

Reply:
Yeah - especially since it's stored as (a pointer to) the interface, makes it trivial here.